sdk/python/kfp/client/auth.py

```diff
@@ -179,40 +190,149 @@ def get_google_open_id_connect_token(service_account_credentials):
+    Args:
+        host (str): hostname of redirect_uri
+        port (int): port of redirect_uri
+        auth_url (str): OAuth authorization code request URL
+    Returns:
+        str: a URL containing authorization code
+    """
     print(auth_url)
-    open_new_tab(auth_url)
-    return input(
-        "If there's no browser window prompt, please direct to the URL above, "
-        'then copy and paste the authorization code here: ')
+    authorization_response = input(
+        'Carefully follow these steps: (1) open the URL above in your'
+        ' browser, (2) authenticate and copy a url of the response page'
+        f' that starts with http://{host}:{port}..., and (3) paste it'
+        ' below:\n')
+    return authorization_response
+
+
+def get_auth_response_local(host: str, port: int, auth_url: str):
+    """Fetches OAuth authorization response URL using a local web-server.
+
+    Args:
+        host (str): hostname of the server
+        port (int): port of the server
+        auth_url (str): OAuth authorization code request URL
+    Returns:
+        str: a URL containing authorization code
+    """
+    with get_local_server_app(host, port) as (local_server, wsgi_app):
+        open_new_tab(auth_url)
+        print(f'Please visit this URL to authorize Kubeflow SDK: {auth_url}.')
+        local_server.handle_request()
+        return wsgi_app.last_request_uri
 
 
-def get_refresh_token_from_code(auth_code, client_id, client_secret):
+def get_refresh_token_from_code(auth_code: str, client_id: str,
+                                client_secret: str, redirect_uri: str):
+    """Returns refresh or access token from authorization code.
+
+    Args:
+        auth_code (str): OAuth authorization code
+        client_id (str): OAuth client ID
+        client_secret (str): OAuth client secret
+        redirect_uri (str): redirect uri used to obtain auth_code
+    Returns:
+        str: long-lived refresh or short-lived access token
+    """
     payload = {
         'code': auth_code,
         'client_id': client_id,
         'client_secret': client_secret,
-        'redirect_uri': 'urn:ietf:wg:oauth:2.0:oob',
+        'redirect_uri': redirect_uri,
         'grant_type': 'authorization_code'
     }
     res = requests.post(OAUTH_TOKEN_URI, data=payload)
-    res.raise_for_status()
-    return str(json.loads(res.text)[u'refresh_token'])
+    try:
+        res.raise_for_status()
+    except requests.exceptions.HTTPError as err:
+        raise ValueError(
+            ('Some HTTPErrors are caused by expired credentials in '
+             '~/.config/kfp/context.json. Try renaming or moving '
+             'them to another directory before trying again.')) from err
+
+    parsed_res = json.loads(res.text)
+    if 'refresh_token' in parsed_res:
+        return str(parsed_res['refresh_token'])
+    return str(parsed_res['access_token'])
```

Comment:
Can we mix the usage of access_token and refresh_token here?

Reply:
I changed it a little as @chensun suggested. I would avoid converting this into a single line in favor of keeping the code more readable. Let me know, if you feel the other way.

Reply:
My point is actually: `get_refresh_token_from_client_id()` shouldn't return access token. It might cause error when access token is used for `refresh_token` field.

Reply:
his is a good point. As I understand, `refresh_token` is only given once, when you authenticate for the first time. After that, you only get `access_token`. The difference is that `refresh_token` get obsolete after 6-months period of inactivity or when you revoke access from that account, while access_token are short-lived. Refresh tokens are used to obtain ID tokens or access tokens and are typically stored in a DB, while access tokens are self-contained and can be decoded to validate access to a particular resource.

There are [a long discussion](https://github.com/googleapis/google-api-python-client/issues/213) and [official documentation](https://developers.google.com/identity/protocols/oauth2/web-server#exchange-authorization-code) with options to enforce `refresh_token`. I tried adding `access_type=offline&prompt=consent` as was suggested there, but it doesn't guarantee `refresh_token` presense in the response. Instead, [other docs](https://developers.google.com/identity/protocols/oauth2/web-server#offline) suggest that there are limits on the number of refresh tokens that can be issued. In my testing, presumably, I reached the limit and I am no longer getting `refresh_token`.

During testing of the proposed change, both scenarios with `access_token` and `refresh_token` were functioning as expected for this particular use case:
```python
client = kfp.Client(host=...,
client_id=...,
other_client_id=...,
other_client_secret=...)
```

So, perhaps, refresh_token and access_token can be used interchangeably? @connor-mccarthy @chensun ?

If yes, I could rename functions to avoid confusion. And store the returned token in `~/.config/kfp/credentials.json` as `refresh_token` or `access_token` whichever is returned. Besides, we might also need to return a boolean flag here [lines 272-264 in client/client.py](https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/client/client.py#L262-L264) depending on what token type is returned. Should we change anything else, @chensun @connor-mccarthy ?

Another option is to raise a `KeyError` if `refresh_token` is missing and instruct users to revoke access manually and repeat.

Please, correct me if I am wrong as my expertise is fairly limited.

Reply:
Sounds good to me for storing `access_token` and `refresh_token` separately. If the `refresh_token` is missing, it makes sense to fall back to the manual approach by recreating a KFP client. It is a great summary you put here!